python_modules/dagster/dagster/_core/definitions/logical_version.py

```diff
@@ -0,0 +1,178 @@
+    )
+):
+    """Class that represents a logical version for an asset.
+
+    Args:
+        value (str): An arbitrary string representing a logical version.
+    """
+
+    def __new__(
+        cls,
+        value: str,
+    ):
+        return super(LogicalVersion, cls).__new__(
+            cls,
+            value=check.str_param(value, "value"),
+        )
+
+
+UNKNOWN_LOGICAL_VERSION: Final[LogicalVersion] = LogicalVersion("UNKNOWN")
+DEFAULT_LOGICAL_VERSION: Final[LogicalVersion] = LogicalVersion("INITIAL")
+
+
+class LogicalVersionProvenance(
+    NamedTuple(
+        "_LogicalVersionProvenance",
+        [
+            ("code_version", str),
+            ("input_logical_versions", Mapping["AssetKey", LogicalVersion]),
+        ],
+    )
+):
+    def __new__(
+        cls,
+        code_version: str,
+        input_logical_versions: Mapping["AssetKey", LogicalVersion],
+    ):
+        from dagster._core.definitions.events import AssetKey
+        return super(LogicalVersionProvenance, cls).__new__(
+            cls,
+            code_version=check.str_param(code_version, "code_version"),
+            input_logical_versions=check.mapping_param(
+                input_logical_versions,
+                "input_logical_versions",
+                key_type=AssetKey,
+                value_type=LogicalVersion,
+            ),
+        )
+
+
+# ########################
+# ##### TAG KEYS
+# ########################
+
+LOGICAL_VERSION_TAG_KEY: Final[str] = "dagster/logical_version"
+CODE_VERSION_TAG_KEY: Final[str] = "dagster/code_version"
+INPUT_LOGICAL_VERSION_TAG_KEY_PREFIX: Final[str] = "dagster/input_logical_version"
+INPUT_EVENT_POINTER_TAG_KEY_PREFIX: Final[str] = "dagster/input_event_pointer"
+
+
+def get_input_logical_version_tag_key(input_key: "AssetKey") -> str:
+    return f"f{INPUT_LOGICAL_VERSION_TAG_KEY_PREFIX}/{input_key.to_user_string()}"
+
+
+def get_input_event_pointer_tag_key(input_key: "AssetKey") -> str:
+    return f"{INPUT_EVENT_POINTER_TAG_KEY_PREFIX}/{input_key.to_user_string()}"
+
+
+# ########################
+# ##### COMPUTE / EXTRACT
+# ########################
+
+
+def compute_logical_version(
+    code_version: Union[str, UnknownValue], input_logical_versions: Sequence[LogicalVersion]
+) -> LogicalVersion:
+    """Compute a logical version from inputs.
+
+    Args:
+        code_version (LogicalVersion): The code version of the computation.
+        input_logical_versions (Sequence[LogicalVersion]): The logical versions of the inputs.
+
+    Returns:
+        LogicalVersion: The computed logical version.
+    """
+    check.inst_param(code_version, "code_version", (str, UnknownValue))
+    check.sequence_param(input_logical_versions, "input_versions", of_type=LogicalVersion)
+
+    if isinstance(code_version, UnknownValue) or UNKNOWN_LOGICAL_VERSION in input_logical_versions:
+        return UNKNOWN_LOGICAL_VERSION
+
+    all_inputs = [code_version, *(v.value for v in input_logical_versions)]
```

Comment:
I don't know how to concisely express that with the `code_version` at the front. `(code_version, *(v.value for v in input_logical_versions))` is a tuple rather than iterable

Reply:
Hmm ok, not a big deal.